File: Code/Java Code/Shebl/src/gp/Leg.java
package gp;

/**
 *
 * @author Abdelrahman
 */
public class Leg {

       public double ID = -1;             //Leg Number, -1 not initialized
       public double Xb = 0;              //The Distance from Origin on the X axis
       public double Yb = 0;              //The Distance form Origin on the Y axis
       public double  Inv = -1;            //Is the Leg flipped (Is it Forward Leg)
       public double  Right = 1;           //Is the Leg on the right side of the robot
       public double[] Theta = {0, 45, 90}; //The Joint's Angles
       public double[] Len = {8, 10.5, 13.1, 1.8}; //Leg Dimentions
       public double[] Swing_Theta;         //The Theta's in the Swing Phase
       public double[] foot;                //The Foot Point
       public double Tup = 0.25;          //Time of rising the leg
       public double Tdown = 0.25;        //Time of downing the leg
       public double Tdisp = 0.5;         //Time of moving the leg
       public double disp = 4;            //The displacement distance
       public double h = 4;
       
       //Constructor
       public Leg(double id, double xb, double yb, double inv, double r)
       {
           ID = id;
           if(inv == 1)
               Xb = xb;
           else
               Xb = -xb;
           
           if(r == 1)
               Yb = -yb;
           else
               Yb = yb;

           Inv = inv;
           Right = r;
           Swing_Theta = new double[6];
           foot = Link3();
       }
       
       //Update Leg
       public void Update(double T0, double T1, double T2)
       {
           if(Theta.length < 3)
           {
               System.out.println("Logical Error in Leg.Update");
               return;
           }
           Theta[0] = T0;
           Theta[1] = T1;
           Theta[2] = T2;
           foot = Link3();
       }
       
       //Move Leg
       public void Move()
       {
           //Get the Velocity of rising the Leg in Z axis
           //Vup = L.h/L.Tup;
           double X = foot[0] - Xb;
           double Z = foot[2] + h;
           if(Inv == 1)
               X = -X;
           
           double T2 = Util.Theta2(X, Z, Theta[0]);
           double T1 = Util.Theta1(X, Z, Theta[0], T2);
           Swing_Theta[0] = T1;
           Swing_Theta[1] = T2;
           //L.Theta = [L.Theta(1), T1, T2];
           Theta[1] = T1;
           Theta[2] = T2;
           foot = Link3();
           
           //Get the Velocity of moving the Leg forward in x axis
           //Vdisp = L.disp/L.Tdisp;
           X = foot[0] - Xb;
           Z = foot[2];
           X = X + disp;
           if(Inv == 1)
               X = -X;
           
           T2 = Util.Theta2(X, Z, Theta[0]);
           T1 = Util.Theta1(X, Z, Theta[0], T2);
           Swing_Theta[2] = T1;
           Swing_Theta[3] = T2;
           //L.Theta = [L.Theta(1), T1, T2];
           Theta[1] = T1;
           Theta[2] = T2;
           foot = Link3();
           
           //Get the Velocity of downing the Leg in z axis
           //Vdown = -L.h/L.Tdown;
           X = foot[0] - Xb;
           Z = foot[2] - h;
           if(Inv == 1)
               X = -X;
           
           T2 = Util.Theta2(X, Z, Theta[0]);
           T1 = Util.Theta1(X, Z, Theta[0], T2);
           Swing_Theta[4] = T1;
           Swing_Theta[5] = T2;
           //L.Theta = [L.Theta(1), T1, T2];
           Theta[1] = T1;
           Theta[2] = T2;
           foot = Link3();
       }
       
       public void Move_Disp(double X, double Y)
       {
           double X_abs = Math.abs(X);
           double Y_abs = Math.abs(Y);
           if((X_abs < Math.pow(10, -10)) && (Y_abs < Math.pow(10, -10)))
               return;
           
           //Get the Req angles to set the foot at (X, Y) from COM
           double X_Old = foot[0] - Xb;
           X = X_Old + X;
           Y = foot[1] - Yb + Y;
           if(Inv == 1)
               X = -X;
           
           if(Right == 1)
               Y = -Y;
           
           double Z = foot[2];
           double T0 = Math.atan(Y/Z);
           double T2 = Util.Theta2(X, Z, T0);
           double T1 = Util.Theta1(X, Z, T0, T2);
           Theta[0] = T0;
           Theta[1] = T1;
           Theta[2] = T2;
           foot = Link3();
       }
       //Third Link
       private double[] Link3()
       {
           double L0 = Len[0];
           double L1 = Len[1];
           double L2 = Len[2];
           double L3 = Len[3];
           double T0,T1,T2;
           if(Right == 1)
               T0 = Theta[0];  //The Leg on the right side
           else
               T0 = Theta[0]; //the Leg on the left side
           
           if(Inv == -1)       //If the robot isn't flipped
           {
               T1 = Theta[1];
               T2 = Theta[2];
           }
           else
           {
               T1 = -Theta[1];
               T2 = -Theta[2];
           }
           
           //Leg Kinematics
           double X = L1*Math.sin(Util.toRad(T1)) + L2*Math.sin(Util.toRad(T1-T2)) + Xb;
           double Temp = L0 + L1*Math.cos(Util.toRad(T1)) + L2*Math.cos(Util.toRad(T1-T2)) + L3;
           double Y = -Math.sin(Util.toRad(T0))*(Temp) + Yb;
           double Z = -Math.cos(Util.toRad(T0))*(Temp);

           double[] p = new double[3];
           p[0] = X;
           p[1] = Y;
           p[2] = Z;
           
           return p;
       }
}

File: Model/Kinematics/Leg.m
classdef Leg
%Class 

   properties
       ID = -1;             %Leg Number, -1 not initialized
       Xb = 0;              %The Distance from Origin on the X axis
       Yb = 0;              %The Distance form Origin on the Y axis
       Inv = -1;            %Is the Leg flipped (Is it Forward Leg)
       Right = 1;           %Is the Leg on the right side of the robot
       Theta = [0, 45, 90]; %The Joint's Angles
       Len = [8, 10.5, 13.1, 1.8];%Leg Dimentions
       foot;                %The Foot Point
       Tup = 0.25;          %Time of rising the leg
       Tdown = 0.25;        %Time of downing the leg
       Tdisp = 0.5;         %Time of moving the leg
       disp = 4;            %The displacement distance
       h = 4;               %The displacement distance
   end

   methods
       %Constructor
       function L = Leg(id, xb, yb, inv, r)
           L.ID = id;
           if(inv == 1)
               L.Xb = xb;
           else
               L.Xb = -xb;
           end
           if(r == 1)
               L.Yb = -yb;
           else
               L.Yb = yb;
           end
           L.Inv = inv;
           L.Right = r;
           L.foot = L.Link3();
       end
       
       %Update Leg
       function L = Update(L, T0, T1, T2)
           L.Theta = [T0, T1, T2];
           L.foot = L.Link3();
       end
       
       %Move Leg
       function L = Move(L)
           %Get the Velocity of rising the Leg in Z axis
           %Vup = L.h/L.Tup;
           X = L.foot(1) - L.Xb;
           Z = L.foot(3) + L.h;
           if(L.Inv == 1)
               X = -X;
           end
           T2 = Theta2(X, Z, L.Theta(1));
           T1 = Theta1(X, Z, L.Theta(1), T2);
           L.Theta = [L.Theta(1), T1, T2];
           L.foot = L.Link3();
           
           %Get the Velocity of moving the Leg forward in x axis
           %Vdisp = L.disp/L.Tdisp;
           X = L.foot(1) - L.Xb;
           Z = L.foot(3);
           X = X + L.disp;
           if(L.Inv == 1)
               X = -X;
           end
           T2 = Theta2(X, Z, L.Theta(1));
           T1 = Theta1(X, Z, L.Theta(1), T2);
           L.Theta = [L.Theta(1), T1, T2];
           L.foot = L.Link3();
           
           %Get the Velocity of downing the Leg in z axis
           %Vdown = -L.h/L.Tdown;
           X = L.foot(1) - L.Xb;
           Z = L.foot(3) - L.h;
           if(L.Inv == 1)
               X = -X;
           end
           T2 = Theta2(X, Z, L.Theta(1));
           T1 = Theta1(X, Z, L.Theta(1), T2);
           L.Theta = [L.Theta(1), T1, T2];
           L.foot = L.Link3();
       end
       
       function L = Move_Disp(L, X, Y)
           if(X == 0 && Y == 0)
               return;
           end
           %Get the Req angles to set the foot at (X, Y) from COM
           X_Old = L.foot(1) - L.Xb;
           X = X_Old + X;
           Y = L.foot(2) - L.Yb + Y;
           if(L.Inv == 1)
               X = -X;
           end
           if(L.Right == 1)
               Y = -Y;
           end
           Z = L.foot(3);
           T0 = atan(Y/Z);
           T2 = Theta2(X, Z, T0);
           T1 = Theta1(X, Z, T0, T2);
           L.Theta = [T0, T1, T2];
           L.foot = L.Link3();
       end
       
       %First Link
       function p = Link0(L)
           p(1) = L.Xb;         %X
           p(2) = L.Yb;         %Y
           p(3) = -L.Len(1);    %Z
       end
       
       %First Link
       function p = Link1(L)
           %Temp Variables
           L0 = L.Len(1);
           L1 = L.Len(2);
           if(L.Right == 1)
               T0 = L.Theta(1);  %The Leg on the right side
           else
               T0 = -L.Theta(1); %the Leg on the left side
           end
           if(L.Inv == -1)       %If the robot isn't flipped
               T1 = L.Theta(2);
           else
               T1 = -L.Theta(2);
           end
           
           %Leg Kinematics
           X = L1*sind(T1) + L.Xb;
           Temp = L0 + L1*cosd(T1);
           Y = -sind(T0)*(Temp) + L.Yb;
           Z = -cosd(T0)*(Temp);
           p = [X, Y, Z];
       end
       
       %Second Link
       function p = Link2(L)
           %Temp Variables
           L0 = L.Len(1);
           L1 = L.Len(2);
           L2 = L.Len(3);
           if(L.Right == 1)
               T0 = L.Theta(1);  %The Leg on the right side
           else
               T0 = -L.Theta(1); %the Leg on the left side
           end
           if(L.Inv == -1)       %If the robot isn't flipped
               T1 = L.Theta(2);
               T2 = L.Theta(3);
           else
               T1 = -L.Theta(2);
               T2 = -L.Theta(3);
           end
           
           %Leg Kinematics
           X = L1*sind(T1) + L2*sind(T1-T2) + L.Xb;
           Temp = L0 + L1*cosd(T1) + L2*cosd(T1-T2);
           Y = -sind(T0)*(Temp) + L.Yb;
           Z = -cosd(T0)*(Temp);
           p = [X, Y, Z];
       end
       
       %Third Link
       function p = Link3(L)
           %Temp Variables
           L0 = L.Len(1);
           L1 = L.Len(2);
           L2 = L.Len(3);
           L3 = L.Len(4);
           if(L.Right == 1)
               T0 = L.Theta(1);  %The Leg on the right side
           else
               T0 = -L.Theta(1); %the Leg on the left side
           end
           if(L.Inv == -1)       %If the robot isn't flipped
               T1 = L.Theta(2);
               T2 = L.Theta(3);
           else
               T1 = -L.Theta(2);
               T2 = -L.Theta(3);
           end
           
           %Leg Kinematics
           X = L1*sind(T1) + L2*sind(T1-T2) + L.Xb;
           Temp = L0 + L1*cosd(T1) + L2*cosd(T1-T2) + L3;
           Y = -sind(T0)*(Temp) + L.Yb;
           Z = -cosd(T0)*(Temp);
           p = [X, Y, Z];
       end
       
       %Draw The Leg
       function Draw(L)
           P0 = L.Link0();
           P1 = L.Link1();
           P2 = L.Link2();
           P3 = L.Link3();
           hold all;
           line('XData',[L.Xb,P0(1)], 'YData',[L.Yb,P0(2)], 'ZData',[0,P0(3)]);
           line('XData',[P0(1),P1(1)], 'YData',[P0(2),P1(2)], 'ZData',[P0(3),P1(3)]);
           line('XData',[P1(1),P2(1)], 'YData',[P1(2),P2(2)], 'ZData',[P1(3),P2(3)]);
           line('XData',[P2(1),P3(1)], 'YData',[P2(2),P3(2)], 'ZData',[P2(3),P3(3)]);
       end
   end
end 
